source/common/common/inline_map.h

```diff
@@ -0,0 +1,459 @@
+        return false;
+      }
+
+      resetInlineMapEntry(*entry_id, std::move(value));
+      return true;
+    } else {
+      // This key is not registered as inline key and try to insert the value to the normal map.
+      return dynamic_entries_.emplace(std::forward<ArgK>(key), std::forward<ArgV>(value)).second;
+    }
+  }
+
+  // Set the entry for the given handle. If the handle is not valid, do nothing.
+  template <class ArgV> bool insert(InlineKey key, ArgV&& value) {
+#ifndef NDEBUG
+    ASSERT(key.descriptorId() == descriptor_.descriptorId(), "Invalid inline key");
+#endif
+
+    // If the entry is already valid, insert will fail and return false.
+    if (inlineEntryValid(key.inlineId())) {
+      return false;
+    }
+
+    resetInlineMapEntry(key.inlineId(), std::forward<ArgV>(value));
+    return true;
+  }
+
+  // Remove the entry for the given key. If the key is not found, do nothing.
+  template <class ArgK> void remove(const ArgK& key) {
+    if (auto entry_id = inlineLookup(key); entry_id.has_value()) {
+      clearInlineMapEntry(*entry_id);
+    } else {
+      dynamic_entries_.erase(key);
+    }
+  }
+
+  // Remove the entry for the given handle. If the handle is not valid, do nothing.
+  void remove(InlineKey key) {
+#ifndef NDEBUG
+    ASSERT(key.descriptorId() == descriptor_.descriptorId(), "Invalid inline key");
+#endif
+    clearInlineMapEntry(key.inlineId());
+  }
+
+  void iterate(std::function<bool(const Key&, const Value&)> callback) const {
+    for (const auto& entry : dynamic_entries_) {
+      if (!callback(entry.first, entry.second)) {
+        return;
+      }
+    }
+
+    for (uint64_t i = 0; i < descriptor_.inlineKeysNum(); ++i) {
+      if (!inlineEntryValid(i)) {
+        continue;
+      }
+
+      if (!callback(descriptor_.inlineKeys()[i], inline_entries_[i])) {
+        return;
+      }
+    }
+  }
+
+  uint64_t size() const { return dynamic_entries_.size() + inline_entries_size_; }
+
+  bool empty() const { return size() == 0; }
+
+  ~InlineMap() {
+    // Destruct all inline entries.
+    for (uint64_t i = 0; i < descriptor_.inlineKeysNum(); ++i) {
+      clearInlineMapEntry(i);
+    }
+    memset(inline_entries_storage_, 0, descriptor_.inlineKeysNum() * sizeof(Value));
+  };
+
+  // Override operator [] to support the normal key assignment.
+  template <class ArgK> Value& operator[](const ArgK& key) {
+    if (auto entry_id = inlineLookup(key); entry_id.has_value()) {
+      // This key is registered as inline key and try to insert the value to the inline array.
+      if (!inlineEntryValid(*entry_id)) {
+        resetInlineMapEntry(*entry_id, Value());
+      }
+      return inline_entries_[*entry_id];
+    } else {
+      // This key is not registered as inline key and try to insert the value to the normal map.
+      return dynamic_entries_[key];
+    }
+  }
+
+  // Override operator [] to support the inline key assignment.
+  Value& operator[](InlineKey key) {
+#ifndef NDEBUG
```

Comment:
Seems ASSERT is controlled by more complex conditon rather than only simple NDEBIG. And because the `descriporId()` method is only compiled when the NEDBUG is not defined, so if remove this macro, the compile will failed.